src/python/pants/jvm/resources.py

```diff
@@ -76,14 +77,21 @@ async def assemble_resources_jar(
     output_filename = f"{request.component.representative.address.path_safe_spec}.resources.jar"
     output_files = [output_filename]
 
+    # #16231: Valid JAR files need the directories of each resource file as well as the files
+    # themselves.
+
+    paths = {Path(filename) for filename in source_files.snapshot.files}
+    directories = {parent for path in paths for parent in path.parents}
+    input_files = {str(path) for path in chain(paths, directories)}
```

Comment:
Should duplicates be removed from directories?

Reply:
`input_files` is a set, so this is already deduplicated.

Reply:
Make sense. Is set safe for caching? @stuhood mentioned sets have undeterministic order in python.

Reply:
In this context, it's safe for caching, since this `Process` rule would only get called if the `JvmResourcesRequest` rule got called, and that would only happen if the inputs changed.

I could use `sorted()` here, which would make the process result cacheable, but I'm not 100% certain that matters here.

Reply:
Cool, thanks 🙏

Reply:
> I could use sorted() here, which would make the process result cacheable, but I'm not 100% certain that matters here.

It does matter: the digest of `Process` structs are our primary cache key, and run over run you would end up re-jar'ing things.

The rule of thumb (which I hope to add a [mypy lint for](https://github.com/pantsbuild/pants/issues/14195#issuecomment-1114087237)) is: never iterate over a `set` (only an `(Frozen)OrderedSet`).

Reply:
I've pushed a fix @stuhood